src/partial.ts

```diff
@@ -0,0 +1,76 @@
+import { purry } from './purry';
+
+type RemovePrefix<
+  T extends ReadonlyArray<unknown>,
+  U extends ReadonlyArray<unknown>,
+> = T extends [...U, ...infer V] ? V : never;
+
+/**
+ * Creates a function that calls `func` with `args` put before the arguments
+ * it receives.
+ *
+ * **Note:** This method doesn't set the `length` property of partially applied
+ * functions.
+ *
+ * @param func the function to wrap
+ * @param args the arguments to put before
+ * @returns a partially applied function
+ *
+ * @signature
+ *    R.partial(args, func)
+ *
+ * @example
+ *    const fn = (x, y, z) => `${x}, ${y}, and ${z}`
+ *    const partialFn = R.partial([1, 2], fn)
+ *    partialFn(3) // => 1, 2, and 3
+ *
+ * @dataFirst
+ * @category Function
+ * @see partialRight
+ */
+export function partial<
+  A extends ReadonlyArray<unknown>,
+  B extends ReadonlyArray<unknown>,
+  C,
+>(args: [...A], func: (...args: B) => C): (...rest: RemovePrefix<B, A>) => C;
```

Comment:
To get a more generic and flexible implementation we can use a function type instead of just the array of it's arguments, this provides typescript more power to infer for itself the types:


```suggestion
export function partial<
  T extends IterableContainer,
  // This is how functions are defined in ReturnType and Paramters by typescript
  F extends (...args: any) => any,
>(data: T, func: F): (...rest: RemovePrefix<Parameters<F>, T>) => ReturnType<F>;
```

Reply:
This is solved by using our `IterableContainer` helper type (i tried it!), I think it has something to do with how typescript infers tuples generically. 


Inspired by:
https://github.com/microsoft/TypeScript/blob/2a0edf7949ad8427e3ef233a4c549ecbe32953e1/src/lib/es2015.promise.d.ts#L21